lib/gc/Transforms/CPUPhysicalRegisterPass.cpp

```diff
@@ -0,0 +1,2978 @@
+  // prepare for loop iterargs
+  GenerateLoopHelper loopHelper(groupId, 0);
+  prepareForLoopArgs(groupId, loopHelper);
+
+  ArrayRef<int64_t> shapes = vectorType.getShape();
+  // generate for loop
+  auto forOp = constructNestedForOp(groupId, rewriter, rewriter.getUnknownLoc(),
+                                    shapes, loopHelper);
+  replaceOpUsersWithForLoopResult(forOp, groupId, loopHelper.nextAnchorResults,
+                                  loopHelper.nextAnchorResultsIdxMap,
+                                  loopHelper.nextAnchorResultOrignalResultMap);
+
+  return forOp;
+}
+
+bool LoopGeneratorImpl::isGroupHasSpecialOperation(const size_t grpIdx) {
+  auto &rdCanonicalizer = getMultiRdCanonicalizers()[grpIdx];
+  auto &tpCanonicalizer = getTransposeCanonicalizers()[grpIdx];
+  auto &shapeCastCanonicalizer = getShapeCastCanonicalizers()[grpIdx];
+  return !rdCanonicalizer.getCandidateOps().empty() or
+         !tpCanonicalizer.getCandidateOps().empty() or
+         !shapeCastCanonicalizer.getCandidateOps().empty();
+}
+
+void LoopGeneratorImpl::generateGroupOpVectorizedIR(const int idx) {
+  auto &grp = getVectorBasedFusion().getOpGroups()[idx];
+  if (grp.empty()) {
+    LDBG("Current operation Group is empty.");
+    return;
+  }
+  // TODO: special operation better fusion
+  if (isGroupHasSpecialOperation(idx))
+    return;
+
+  VectorType groupType =
+      getVectorBasedFusion().getGroupBiggestRankVectorType()[idx];
+  IRRewriter rewriter(grp.back());
+  rewriter.setInsertionPointAfter(grp.back());
+  // 1. Rewrite operation as vectorized form
+  // 2. Generate loop
+  // rewriteOperationAsVectorize(rewriter, idx);
+  auto forOp = generateVectorizedForLoop(idx, rewriter, groupType);
+  // special operation do not need to change anything
+  if (failed(forOp))
+    return;
+
+  moveLoopInvariantCode(forOp.value());
+}
+
+/// Pass that lower to physical vector.
+struct CPUPhysicalRegisterPass
+    : public impl::CPUPhysicalRegisterPassBase<CPUPhysicalRegisterPass> {
+
+  void runOnOperation() final {
+    //
+    auto *ctx = &getContext();
+    RewritePatternSet patterns(ctx);
+    auto func = getOperation();
+
+    if (hasNotSupportOperation(&func)) {
+      LDBG("Not support operation appears in current function.");
+      return;
+    }
+    // affineApply operation is always used by other operations.
+    std::function<bool(Operation *)> candidateFunc = isProducerOp;
+    moveSomeInterferenceOperation(&func, ctx, candidateFunc);
+    candidateFunc = isCandidateMoveOperations;
+    moveSomeInterferenceOperation(&func, ctx, candidateFunc);
+    // canonicalize vector operation, default use vector-based fusion
+    // strategy.
+    HardWareInfo hwInfo;
+    CPUTargetDescriptionAnalysis sysDesc =
+        getAnalysis<CPUTargetDescriptionAnalysis>();
+    hwInfo.favx512f = sysDesc.getMaxVectorWidth() >= 512;
+    hwInfo.favx2 = sysDesc.getMaxVectorWidth() >= 256;
```

Comment:
Is it more accurate to define and use `hwInfo.vectorWidth` as it's acquired by the target info by the vector lanes? vectorWidth (256) doesn't mean it's avx2, it could be avx too.

Reply:
Yes~ It should be this!